File: Samples/CS/CodeSamples/ConcatenationSample.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Inetlab.SMPP;
using Inetlab.SMPP.Common;
using Inetlab.SMPP.Headers;
using Inetlab.SMPP.Parameters;
using Inetlab.SMPP.PDU;

namespace CodeSamples
{
    public  class ConcatenationSample
    {
        private readonly SmppConfig _config = new SmppConfig();
        private readonly SmppClient _client = new SmppClient();

        // <SendConcatenatedMessageInUDH>
        public async Task SendConcatenatedMessageInUDH(TextMessage message)
        {
            var builder = SMS.ForSubmit()
                .From(_config.ShortCode, AddressTON.NetworkSpecific, AddressNPI.Unknown)
                .To(message.PhoneNumber)
                .Text(message.Text);

            var resp = await _client.Submit(builder);
        }
        // </SendConcatenatedMessageInUDH>

        // <CreateSumbitSmWithConcatenationInUDH>
        public SubmitSm CreateSumbitSmWithConcatenationInUDH(ushort referenceNumber, byte totalParts, byte partNumber, string textSegment)
        {

            SubmitSm sm = new SubmitSm();
            sm.SourceAddress = new SmeAddress("1111");
            sm.DestinationAddress = new SmeAddress("79171234567");
            sm.DataCoding = DataCodings.Default;
            sm.RegisteredDelivery = 1;
            sm.UserData.ShortMessage = _client.EncodingMapper.GetMessageBytes(textSegment, sm.DataCoding);

            sm.UserData.Headers.Add(new ConcatenatedShortMessage16bit(referenceNumber, totalParts, partNumber));

            return sm;
        }
        // </CreateSumbitSmWithConcatenationInUDH>

        public async Task SendConcatenatedMessageWithSARParameters(TextMessage message)
        {
            // <ConcatenationWithSARParameters>
            var builder = SMS.ForSubmit()
                .From(_config.ShortCode, AddressTON.NetworkSpecific, AddressNPI.Unknown)
                .To(message.PhoneNumber)
                .Text(message.Text);

            builder.ConcatenationInSAR();

            var resp = await _client.Submit(builder);

            // </ConcatenationWithSARParameters>
        }

        public async Task SendConcatenatedMessageInMessagePayload(TextMessage message)
        {
            // <ConcatenationInMessagePayload>
            var builder = SMS.ForSubmit()
                .From(_config.ShortCode, AddressTON.NetworkSpecific, AddressNPI.Unknown)
                .To(message.PhoneNumber)
                .Text(message.Text);

            builder.MessageInPayload();

            var resp = await _client.Submit(builder);

            // </ConcatenationInMessagePayload>
        }

        // <GetConcatenationFromUDH>
        public Concatenation GetConcatenationFromUDH(SubmitSm data)
        {
            ConcatenatedShortMessages8bit udh8 = data.UserData.Headers.Of<ConcatenatedShortMessages8bit>().FirstOrDefault();

            if (udh8 == null) return null;

            return new Concatenation(udh8.ReferenceNumber, udh8.Total, udh8.SequenceNumber);
        }
        // </GetConcatenationFromUDH>


        // <GetConcatenationFromOptions>
        public Concatenation GetConcatenationFromTLVOptions(SubmitSm data)
        {
            ushort refNumber = 0;
            byte total = 0;
            byte seqNum = 0;

            var referenceNumber = data.Parameters.Of<SARReferenceNumberParamter>().FirstOrDefault();
            if (referenceNumber != null)
            {
                refNumber = referenceNumber.ReferenceNumber;
            }
            var totalSegments = data.Parameters.Of<SARTotalSegmentsParameter>().FirstOrDefault();
            if (totalSegments != null)
            {
                total = totalSegments.TotalSegments;
            }
            var sequenceNumber = data.Parameters.Of<SARSequenceNumberParameter>().FirstOrDefault();
            if (sequenceNumber != null)
            {
                seqNum = sequenceNumber.SequenceNumber;
            }

            return new Concatenation(refNumber, total, seqNum);
        }
        // </GetConcatenationFromOptions>

        public class TextMessage
        {
            public string Id { get; set; }
            public string PhoneNumber { get; set; }
            public string Text { get; set; }

            public string ServiceAddress { get; set; }
        }
    }



    public class SmppConfig
    {
        public string ShortCode { get; set; }
    }
}

File: Samples/VB.NET/CodeExamples/ConcatenationSample.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports Inetlab.SMPP
Imports Inetlab.SMPP.Common
Imports Inetlab.SMPP.Headers
Imports Inetlab.SMPP.Parameters
Imports Inetlab.SMPP.PDU

Namespace CodeSamples
	Public Class ConcatenationSample
		Private ReadOnly _config As New SmppConfig()
		Private ReadOnly _client As New SmppClient()

		' <SendConcatenatedMessageInUDH>
		Public Async Function SendConcatenatedMessageInUDH(ByVal message As TextMessage) As Task
			Dim builder = SMS.ForSubmit().From(_config.ShortCode, AddressTON.NetworkSpecific, AddressNPI.Unknown).To(message.PhoneNumber).Text(message.Text)

			Dim resp = Await _client.Submit(builder)
		End Function
		' </SendConcatenatedMessageInUDH>

		' <CreateSumbitSmWithConcatenationInUDH>
		Public Function CreateSumbitSmWithConcatenationInUDH(ByVal referenceNumber As UShort, ByVal totalParts As Byte, ByVal partNumber As Byte, ByVal textSegment As String) As SubmitSm

			Dim sm As New SubmitSm()
			sm.SourceAddress = New SmeAddress("1111")
			sm.DestinationAddress = New SmeAddress("79171234567")
			sm.DataCoding = DataCodings.Default
			sm.RegisteredDelivery = 1
			sm.UserData.ShortMessage = _client.EncodingMapper.GetMessageBytes(textSegment, sm.DataCoding)

			sm.UserData.Headers.Add(New ConcatenatedShortMessage16bit(referenceNumber, totalParts, partNumber))

			Return sm
		End Function
		' </CreateSumbitSmWithConcatenationInUDH>

		Public Async Function SendConcatenatedMessageWithSARParameters(ByVal message As TextMessage) As Task
			' <ConcatenationWithSARParameters>
			Dim builder = SMS.ForSubmit().From(_config.ShortCode, AddressTON.NetworkSpecific, AddressNPI.Unknown).To(message.PhoneNumber).Text(message.Text)

			builder.ConcatenationInSAR()

			Dim resp = Await _client.Submit(builder)

			' </ConcatenationWithSARParameters>
		End Function

		Public Async Function SendConcatenatedMessageInMessagePayload(ByVal message As TextMessage) As Task
			' <ConcatenationInMessagePayload>
			Dim builder = SMS.ForSubmit().From(_config.ShortCode, AddressTON.NetworkSpecific, AddressNPI.Unknown).To(message.PhoneNumber).Text(message.Text)

			builder.MessageInPayload()

			Dim resp = Await _client.Submit(builder)

			' </ConcatenationInMessagePayload>
		End Function

		' <GetConcatenationFromUDH>
		Public Function GetConcatenationFromUDH(ByVal data As SubmitSm) As Concatenation
			Dim udh8 As ConcatenatedShortMessages8bit = data.UserData.Headers.Of(Of ConcatenatedShortMessages8bit)().FirstOrDefault()

			If udh8 Is Nothing Then
				Return Nothing
			End If

			Return New Concatenation(udh8.ReferenceNumber, udh8.Total, udh8.SequenceNumber)
		End Function
		' </GetConcatenationFromUDH>


		' <GetConcatenationFromOptions>
		Public Function GetConcatenationFromTLVOptions(ByVal data As SubmitSm) As Concatenation
			Dim refNumber As UShort = 0
			Dim total As Byte = 0
			Dim seqNum As Byte = 0

			Dim referenceNumber = data.Parameters.Of(Of SARReferenceNumberParamter)().FirstOrDefault()
			If referenceNumber IsNot Nothing Then
				refNumber = referenceNumber.ReferenceNumber
			End If
			Dim totalSegments = data.Parameters.Of(Of SARTotalSegmentsParameter)().FirstOrDefault()
			If totalSegments IsNot Nothing Then
				total = totalSegments.TotalSegments
			End If
			Dim sequenceNumber = data.Parameters.Of(Of SARSequenceNumberParameter)().FirstOrDefault()
			If sequenceNumber IsNot Nothing Then
				seqNum = sequenceNumber.SequenceNumber
			End If

			Return New Concatenation(refNumber, total, seqNum)
		End Function
		' </GetConcatenationFromOptions>

	End Class

	Public Class TextMessage
		Public Property Id() As String
		Public Property PhoneNumber() As String
		Public Property Text() As String

		Public Property ServiceAddress() As String
	End Class

	Public Class SmppConfig
		Public Property ShortCode() As String
	End Class
End Namespace
